site/_posts/2021-09-16-how-to-authenticate-your-react-app.md

```diff
@@ -0,0 +1,1087 @@
+
+But why specify this? Because without specifying where to send the app after authenticating, a bad actor could put in their own redirect URL, sending the browser to their malicious server and gaining access to a token that could be used to view resources as a user. That would be no good!
+
+### Exchange the Authorization Code for an Access Token
+
+An Authorization Code isn’t enough to access the user’s resources, though. For that, we need an Access Token. This is standard OAuth, not something unique to FusionAuth. This step is called the Code Exchange, because we send the auth code to FusionAuth’s `/token` endpoint and receive an Access Token in exchange.  Then that Access Token is passed to the resource server in exchange for the desired resources.  
+
+Add the `/oauth-callback` route file:
+
+```bash
+touch server/routes/oauth-callback.js
+```
+
+For reference, your file tree in `server` should look like this:
+
+```
+server
+├─routes
+│ ├─login.js
+│ ├─oauth-callback.js*
+│ └─user.js
+└─index.js
+```
+
+In `server.js/index.js`...
+
+```js
+// ...
+const oauthCallbackRoute = (require('./routes/oauth-callback'));
+
+// use routes
+app.use('/user', require('./routes/user'));
+app.use('/login', require('./routes/login'));
+app.use('/oauth-callback', require('./routes/oauth-callback'));
+
+// ...
+```
+
+`/oauth-callback` receives an Authorization Code from FusionAuth as a URL parameter during the redirect. If you inspect your browser’s location, you’ll see that there’s a code parameter in the URL:
+
+```
+http://localhost:3001/oauth-callback?
+  code=14CJG_fDl31E5U-1VHOBedsLESZQr3sgt63BcVrGoTU&
+  locale=en_US&
+  userState=Authenticated
+```
+
+Make a new file in `server/routes` called  `oauth-callback.js` and put in the following:
+
+The exchange takes the form of a POST request:
+
+```js
+
+const express = require('express');
+const router = express.Router();
+const request = require('request');
+const config = require('../../config');
+
+router.get('/', (req, res) => {
+  request(
+    // POST request to /token endpoint
+    {
+      method: 'POST',
+      uri: `http://localhost:${config.fusionAuthPort}/oauth2/token`,
+      form: {
+        'client_id': config.clientID,
+        'client_secret': config.clientSecret,
+        'code': req.query.code,
+        'grant_type': 'authorization_code',
+        'redirect_uri': config.redirectURI
+      }
+    },
+
+    // callback lambda
+    (error, response, body) => {
+      // save token to session
+      req.session.token = JSON.parse(body).access_token;
+
+      // redirect to the React app
+      res.redirect(`http://localhost:${config.clientPort}`);
+    }
+  );
+});
+
+module.exports = router;
+```
+
+There are two important things to call out here:
+
+- The `redirect_uri` which is the express server route that you configured as an authorized redirect URI initially. This is present because we track it throughout the process. If it ever differs from what we expect, it's an error condition.
```

Comment:
```suggestion
- The `redirect_uri` which is the express server route that you configured as an authorized redirect URI in the FusionAuth application configuration. This is present because we track it throughout the process. If it ever differs from what we expect, it's an error condition.
```